Core: Fix sidebar accessibility order for screen readers

yatishgoel opened the pull request:
Closes #31076

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

This PR addresses accessibility issue #31076 where the screen reader focus order did not match the visual layout, particularly for the sidebar navigation.

1.  **Desktop View:** Changed the DOM order in `Layout.tsx` to render the `<SidebarContainer>` before the main content (`<MainContentMatcher>`), ensuring screen readers announce the sidebar first.
2.  **Mobile View:** Based on review feedback, also adjusted the mobile layout in `Layout.tsx`. Moved the `<MobileNavigation>` component to render first in the DOM, but used CSS `order: 1` (via a wrapper `div`) to keep it visually positioned at the bottom. This ensures the correct reading order for screen readers on mobile as well.

**TL;DR:** This PR initially fixed the sidebar/content DOM order for desktop screen reader accessibility. Based on review feedback, it now also corrects the DOM order for the mobile navigation (placing it first in the DOM but visually last using CSS `order`), addressing the core reading order issue on both platforms. Other mobile navigation accessibility improvements (markup, focus trapping) will be handled in separate follow-up work.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

This PR modifies the Storybook layout structure to improve accessibility by adjusting the DOM order of sidebar and main content components for screen readers.

- Modified `code/core/src/manager/components/layout/Layout.tsx` to render `SidebarContainer` before `Match` and `ContentContainer` components
- Change ensures screen readers encounter sidebar content before main content, matching the visual left-to-right layout
- Maintains existing responsive behavior and functionality while improving accessibility
- Addresses user feedback from issue #31076 about screen reader content order not matching visual layout

The changes are focused and minimal, only adjusting component render order to improve accessibility without affecting the visual presentation or core functionality.



<!-- /greptile_comment -->

yatishgoel committed:
fix(a11y): ensure sidebar precedes main content in DOM for screen readers

yatishgoel committed:
refactor(layout): extract MainContentMatcher component for improved readability

Sidnioulz reviewed (changes requested):
Manually tested this and it fixes the issue in the Desktop layout.  

It doesn't help the issue wrt. mobile nav, however. For mobile, we'd need MobileNavigation to be first in the DOM order, and its rendered location adjusted however necessary with `order` or CSS positioning.

The mobile navigation button HTML markup is also insufficient. At the very least, we would need
* an aria-label instead of title
* an `aria-expanded` or `aria-controls` markup to indicate the button opens a menu, with the accompanying changes to the menu's markup (e.g. `listbox` or `dialog` role); or the new popover API could maybe be used (e.g. https://mdn.github.io/dom-examples/popover-api/nested-popovers/) The exact markup to use must be evidence-based (e.g. `a11ysupport` analysis, copying an APG, following guidelines from a reputable source or providing a screenreader outcome)
* the keyboard focus order to be fixed and a focus trap to be installed on the menu

We should also change the current HTML markup, so that we have the `nav` element rendered in the DOM but invisible until focused / until the button is triggered. This would help with screenreader keyboard shortcuts that navigate through nav elements or HTML landmarks.

Some references used to write this review:
* https://mdn.github.io/dom-examples/popover-api/nested-popovers
* https://accessibleweb.dev/navigation
* https://theadminbar.com/accessibility-weekly/mobile-nav-and-hamburger-menus/
* https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-navigation/

@shilman I feel there's so much more that needs to be done to fix the mobile menu it may be worth opening a separate issue, or merging this PR but keeping the issue opened with an up-to-date description. What would be your preferred way to proceed?

yatishgoel commented:
> Manually tested this and it fixes the issue in the Desktop layout.
> 
> It doesn't help the issue wrt. mobile nav, however. For mobile, we'd need MobileNavigation to be first in the DOM order, and its rendered location adjusted however necessary with `order` or CSS positioning.
> 
> The mobile navigation button HTML markup is also insufficient. At the very least, we would need
> 
> * an aria-label instead of title
> * an `aria-expanded` or `aria-controls` markup to indicate the button opens a menu, with the accompanying changes to the menu's markup (e.g. `listbox` or `dialog` role); or the new popover API could maybe be used (e.g. https://mdn.github.io/dom-examples/popover-api/nested-popovers/) The exact markup to use must be evidence-based (e.g. `a11ysupport` analysis, copying an APG, following guidelines from a reputable source or providing a screenreader outcome)
> * the keyboard focus order to be fixed and a focus trap to be installed on the menu
> 
> We should also change the current HTML markup, so that we have the `nav` element rendered in the DOM but invisible until focused / until the button is triggered. This would help with screenreader keyboard shortcuts that navigate through nav elements or HTML landmarks.
> 
> Some references used to write this review:
> 
> * https://mdn.github.io/dom-examples/popover-api/nested-popovers
> * https://accessibleweb.dev/navigation
> * https://theadminbar.com/accessibility-weekly/mobile-nav-and-hamburger-menus/
> * https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-navigation/
> 
> @shilman I feel there's so much more that needs to be done to fix the mobile menu it may be worth opening a separate issue, or merging this PR but keeping the issue opened with an up-to-date description. What would be your preferred way to proceed?

Hi @Sidnioulz,

Thanks again for the detailed review and testing! I appreciate you confirming the desktop fix and outlining the remaining mobile-view issues.

I understand my current changes only address the desktop DOM order. To tackle the mobile DOM order (per your note:  
> “For mobile, we’d need `MobileNavigation` to be first in the DOM order, and its rendered location adjusted however necessary with order or CSS positioning.”)  

I propose:

1. **Layout change**  
   In `Layout.tsx`, inside the `isMobile` block, move `<MobileNavigation />` so it renders before `<MainContentMatcher />`.

2. **Visual order**  
   Since `LayoutContainer` uses `flex-direction: column` on mobile, apply `order: 1` to the `MobileNavigation` component (or its wrapper). The other elements will default to `order: 0`, keeping `MobileNavigation` visually at the bottom.

Let me know if this approach for fixing just the mobile DOM order in this PR works for you, or if you’d rather adjust it further here. Thanks again for all your feedback!

Sidnioulz commented:
> I propose:
> 
> 1. **Layout change**
>    In `Layout.tsx`, inside the `isMobile` block, move `<MobileNavigation />` so it renders before `<MainContentMatcher />`.
> 2. **Visual order**
>    Since `LayoutContainer` uses `flex-direction: column` on mobile, apply `order: 1` to the `MobileNavigation` component (or its wrapper). The other elements will default to `order: 0`, keeping `MobileNavigation` visually at the bottom.
> 
> Let me know if this approach for fixing just the mobile DOM order in this PR works for you, or if you’d rather adjust it further here. Thanks again for all your feedback!

That sounds great to me! It would indeed address the DOM ordering in full, and would allow us to focus on improving the markup in a separate issue. Let's see if @shilman confirms this would work.

yatishgoel committed:
Merge branch 'next' into fix/a11y-sidebar-dom-order

shilman commented:
SGTM guys! @ghengeveld  do you have any concerns with this proposal?

ghengeveld commented:
No objections. PR needs a description though. @yatishgoel could you add a few lines on what you did to resolve the issue in the PR description? And perhaps a TLDR on the desktop vs mobile discussion?

yatishgoel committed:
fix(a11y): Adjust mobile nav DOM order for screen readers\n\nFollows up on PR feedback to place mobile navigation first in DOM\nwhile preserving visual layout using CSS order.

Sidnioulz commented on code/core/src/manager/components/layout/Layout.tsx:
```diff
@@ -175,14 +182,18 @@ export const Layout = ({ managerLayoutState, setManagerLayoutState, hasTab, ...s
+
       {isMobile && (
         <>
+          <div style={{ order: 1 }}>
```
nit: For consistency's sake, it would be better to create a `styled(MobileNavigation)` with that `order` property.

yatishgoel replied:
Applied the suggestion, thanks!

Sidnioulz reviewed (commented):
LGTM otherall. Only small issue: DOM layout could be simplified by making full use of `styled` and avoiding an extra `div` layer.

yatishgoel committed:
refactor(layout): Create OrderedMobileNavigation component for improved structure
Replaces inline styling for mobile navigation with a dedicated styled component.

yatishgoel committed:
Merge branch 'next' into fix/a11y-sidebar-dom-order

ghengeveld commented:
@yatishgoel There's a couple of visual regressions [reported by Chromatic](https://www.chromatic.com/build?appId=635781f3500dd2c49e189caf&number=19043). The mobile menu bar seems to be at the top now, at least in these stories. Probably the stories' render functions need to be updated.

yatishgoel committed:
Merge branch 'next' into fix/a11y-sidebar-dom-order

ghengeveld commented:
@yatishgoel That's probably fine as long as it renders as it always has.

yatishgoel committed:
Refactor: Simplify MobileNavigation story layout by removing unnecessary div wrappers

ghengeveld committed:
Revert "Refactor: Simplify MobileNavigation story layout by removing unnecessary div wrappers"

This reverts commit ee47e4041b8a060cf3c839a739473bcf25c1a2b5.

ghengeveld committed:
Pass order prop down to container

ghengeveld committed:
Merge branch 'next' into fix/a11y-sidebar-dom-order

ghengeveld commented:
@yatishgoel Your last commit didn't fix it, in fact it broke something else, so I reverted that change. I took a look at the problem and found that you forgot to properly pass the `order` style down to `MobileNavigation` correctly. Wrapping it as a styled component didn't work because it didn't pass the `className` prop down to the container element, so I fixed that.
